### PR TITLE
feat(cli): disable/enable all <report-type> compliance command

### PR DESCRIPTION
**Pull request by dmurray-lacework**
## Summary

Add 3 new experimental commands 
`lacework compliance aws|azure|gcp disable-report <report_type>`
`lacework compliance aws|azure|gcp enable-report <report_type>`
`lacework compliance aws|azure|gcp report-status <report_type>`

`lacework compliance aws|azure|gcp report-status <report_type> --json`
The json output will return the current state of recommendations for the specified report in a format compatible with the recommendations patch request. It can be used to "revert" state back after a bulk disable if needed.

It is recommended to run `lacework compliance aws|azure|gcp report-status <report_type> --json`
first before bulk disable.

*note report state is cached with exp of 30mins after successful read and patch
run with --nocache flag to ignore cache

#### Supported Report Types
aws: `AWS_CIS_S3`
azure: `AZURE_CIS` `AZURE_CIS_131`
gcp: `GCP_CIS` `GCP_CIS12`

## How did you test this change?
Run manually:
`lacework compliance aws|azure|gcp disable-report <report_type>`
`lacework compliance aws|azure|gcp enable-report <report_type>`
`lacework compliance aws|azure|gcp report-status <report_type>`

New unit tests:
`api/v1_recommendations_test.go`
`api/v1_recommendations_aws_test.go`
`api/v1_recommendations_azure_test.go`
`api/v1_recommendations_gcp_test.go`

## Issue

https://lacework.atlassian.net/browse/ALLY-934
https://lacework.atlassian.net/browse/ALLY-935

### CODE REVIEW

File: api/compliance_gcp.go [hunk 1]
```diff
@@ -38,7 +38,7 @@ func (svc *ComplianceService) GetGcpReport(config ComplianceGcpReportConfig) (
 	err error,
 ) {
 	if config.OrganizationID == "" || config.ProjectID == "" {
-		err = errors.New("organization_id and project_id is required")
+		err = errors.New("organization_id and project_are required")
```
Comment (afiune) [inline api/compliance_gcp.go:41]: ?

File: api/compliance_gcp.go [hunk 2]
```diff
@@ -55,7 +55,7 @@ func (svc *ComplianceService) GetGcpReport(config ComplianceGcpReportConfig) (
 
 func (svc *ComplianceService) DownloadGcpReportPDF(filepath string, config ComplianceGcpReportConfig) error {
 	if config.OrganizationID == "" || config.ProjectID == "" {
-		return errors.New("organization_id and project_id is required")
+		return errors.New("organization_id and project_id are required")
```
Comment (afiune) [inline api/compliance_gcp.go:58]: Thank you 💛

File: api/v1_recommendations_azure.go [hunk 3]
```diff
@@ -31,3 +31,7 @@ func (svc *AzureRecommendationsV1) List() ([]RecommendationV1, error) {
 func (svc *AzureRecommendationsV1) Patch(recommendations RecommendationStateV1) (response RecommendationResponseV1, err error) {
 	return svc.client.Recommendations.patch(AzureRecommendation, recommendations)
 }
+
+func (svc *AzureRecommendationsV1) GetReport(reportType string) (response []RecommendationV1, err error) {
+	return []RecommendationV1{}, nil
```
Comment (afiune) [inline api/v1_recommendations_azure.go:36]: Are we implementing this?

File: api/v1_recommendations_gcp.go [hunk 4]
```diff
@@ -31,3 +33,54 @@ func (svc *GcpRecommendationsV1) List() ([]RecommendationV1, error) {
...
+	}
+
+	var projectID = gcpCfg.Data[0].Data.ID
+	var orgID = "n/a"
+
+	// TODO get all gcp projects for org
+	// cli/cmd/compliance_gcp.go:496
+	//if gcpCfg.Data[0].Data.IDType == "ORGANIZATION" {
+	//
+	//}
```
Comment (afiune) [inline api/v1_recommendations_gcp.go:57]: what's up with this commented code and TODO?

File: api/v1_recommendations_gcp.go [hunk 5]
```diff
@@ -31,3 +33,54 @@ func (svc *GcpRecommendationsV1) List() ([]RecommendationV1, error) {
 func (svc *GcpRecommendationsV1) Patch(recommendations RecommendationStateV1) (response RecommendationResponseV1, err error) {
 	return svc.client.Recommendations.patch(GcpRecommendation, recommendations)
 }
+
+// GetReport This is an experimental feature. Returned RecommendationID's are not guaranteed to be correct. Scoped to Lacework Account/Subaccount
+func (svc *GcpRecommendationsV1) GetReport(reportType string) (response []RecommendationV1, err error) {
+	gcpCfg, err := svc.client.Integrations.ListGcpCfg()
+	if err != nil {
+		return []RecommendationV1{}, err
```
Comment (afiune) [inline api/v1_recommendations_gcp.go:41]: If you are using named returns you have to use naked returns (somehow)
```suggestion
		return
```

Comment (afiune) [reply api/v1_recommendations_gcp.go:41]: Or don't use named returns 😅

File: cli/cmd/compliance_aws.go [hunk 6]
```diff
@@ -243,6 +245,91 @@ To show recommendation details and affected resources for a recommendation id:
...
+		Aliases: []string{"status"},
+		PreRunE: func(_ *cobra.Command, args []string) error {
+			switch args[0] {
+			case "CIS":
+				args[0] = fmt.Sprintf("AWS_%s_S3", args[0])
+				return nil
+			case "AWS_CIS_S3":
+				return nil
+			default:
+				return errors.New("CIS is the only supported report type")
```
Comment (afiune) [inline cli/cmd/compliance_aws.go:300]: Is this coming from Product? or how do we know CIS is the only supported report type?

File: cli/cmd/compliance_aws.go [hunk 7]
```diff
@@ -251,6 +338,10 @@ func init() {
 	complianceAwsCmd.AddCommand(complianceAwsListAccountsCmd)
 	complianceAwsCmd.AddCommand(complianceAwsRunAssessmentCmd)
 
+	// Experimental Commands
```
Comment (afiune) [inline cli/cmd/compliance_aws.go:341]: These commands are marked as experimental (in this comment) but not hidden from help messages.

File: api/v1_recommendations_azure.go (hunk 3 above)
Comment (dmurray-lacework) [reply api/v1_recommendations_azure.go:36]: yes, azure will also be implemented

File: api/compliance_gcp.go (hunk 1 above)
Comment (dmurray-lacework) [reply api/compliance_gcp.go:41]: ugh, messed up correction. See below for what it should be

File: cli/cmd/compliance_aws.go (hunk 7 above)
Comment (dmurray-lacework) [reply cli/cmd/compliance_aws.go:341]: 👍  will mark as hidden

File: api/v1_recommendations_gcp.go (hunk 4 above)
Comment (dmurray-lacework) [reply api/v1_recommendations_gcp.go]: Will be removed be this Draft is moved to a full PR

File: cli/cmd/compliance_aws.go (hunk 6 above)
Comment (dmurray-lacework) [reply cli/cmd/compliance_aws.go:302]: For AWS we are only looking to bulk disable CIS report types. This is in the ticket description written by @marcosgm 
https://lacework.atlassian.net/browse/ALLY-930

File: cli/cmd/compliance_aws.go [hunk 8]
```diff
@@ -243,6 +245,96 @@ To show recommendation details and affected resources for a recommendation id:
...
+			default:
+				return errors.New("CIS is the only supported report type")
+			}
+		},
+		Args: cobra.ExactArgs(1),
+		RunE: func(_ *cobra.Command, args []string) error {
+
+			schema, err := fetchCachedAwsComplianceReportSchema(args[0])
+			if err != nil {
+				return errors.Wrap(err, "unable to aws azure compliance report schema")
```
Comment (rmoles) [inline cli/cmd/compliance_aws.go:271]: Should this say `get aws`?
```suggestion
				return errors.Wrap(err, "unable to get aws compliance report schema")
```

File: cli/cmd/compliance_aws.go [hunk 9]
```diff
@@ -243,6 +245,96 @@ To show recommendation details and affected resources for a recommendation id:
...
+			default:
+				return errors.New("CIS is the only supported report type")
+			}
+		},
+		Args: cobra.ExactArgs(1),
+		RunE: func(_ *cobra.Command, args []string) error {
+			var rows [][]string
+			report, err := fetchCachedAwsComplianceReportSchema(args[0])
+			if err != nil {
+				return errors.Wrap(err, "unable to aws azure compliance report schema")
```
Comment (rmoles) [inline cli/cmd/compliance_aws.go:310]: same as before

File: cli/cmd/compliance_aws.go (hunk 8 above)
Comment (dmurray-lacework) [reply cli/cmd/compliance_aws.go:271]: it should, thanks

File: cli/cmd/compliance_aws.go [hunk 10]
```diff
@@ -285,6 +285,45 @@ To show recommendation details and affected resources for a recommendation id:
...
+	complianceAwsEnableReportCmd = &cobra.Command{
+		Use:     "enable-report <report_type>",
+		Aliases: []string{"enable"},
+		Hidden:  true,
+		PreRunE: func(_ *cobra.Command, args []string) error {
+			switch args[0] {
+			case "CIS":
+				args[0] = fmt.Sprintf("AWS_%s_S3", args[0])
+				return nil
+			case "AWS_CIS_S3":
```
Comment (rmoles) [inline cli/cmd/compliance_aws.go:299]: Is this strictly true? Should it be LW_AWS_S3 of AWS_LW_S3?

File: cli/cmd/compliance_aws.go [hunk 11]
```diff
@@ -243,6 +245,135 @@ To show recommendation details and affected resources for a recommendation id:
...
+		RunE: func(_ *cobra.Command, args []string) error {
+
+			schema, err := fetchCachedAwsComplianceReportSchema(args[0])
+			if err != nil {
+				return errors.Wrap(err, "unable to get aws compliance report schema")
+			}
+
+			// set state of all recommendations in this report to enabled
+			patchReq := api.NewRecommendationV1State(schema, true)
+			response, err := cli.LwApi.Recommendations.Aws.Patch(patchReq)
```
Comment (afiune) [inline cli/cmd/compliance_aws.go:315]: Add progress to this API call. (and all other similar calls)

File: cli/cmd/compliance_aws.go [hunk 12]
```diff
@@ -243,6 +245,135 @@ To show recommendation details and affected resources for a recommendation id:
...
+			// set state of all recommendations in this report to disabled
+			patchReq := api.NewRecommendationV1State(schema, false)
+			response, err := cli.LwApi.Recommendations.Aws.Patch(patchReq)
+			if err != nil {
+				return errors.Wrap(err, "unable to patch aws recommendations")
+			}
+
+			var cacheKey = fmt.Sprintf("compliance/aws/schema/%s", args[0])
+			cli.WriteAssetToCache(cacheKey, time.Now().Add(time.Minute*30), response.RecommendationList())
+			cli.OutputHuman(fmt.Sprintf("All recommendations for report %s have been disabled\n", args[0]))
```
Comment (afiune) [inline cli/cmd/compliance_aws.go:283]: You don't need to concatenate the string, this function already works similar to `fmt.Sprintf()`

Fix all other statements like this one, please.
```suggestion
			cli.OutputHuman("All recommendations for report %s have been disabled\n", args[0])
```

File: cli/cmd/compliance_aws.go [hunk 13]
```diff
@@ -243,6 +245,135 @@ To show recommendation details and affected resources for a recommendation id:
 			return nil
 		},
 	}
+
+	// complianceAwsDisableReportCmd represents the disable-report sub-command inside the aws command
+	// experimental feature
+	complianceAwsDisableReportCmd = &cobra.Command{
+		Use:     "disable-report <report_type>",
+		Aliases: []string{"disable"},
+		Hidden:  true,
```
Comment (afiune) [inline cli/cmd/compliance_aws.go:254]: Even though these are all hidden commands, please add `Short` and `Long` descriptions. If you add them, they won't get documented, but they will be lost on what these commands do if someone finds them.

**Review by afiune — APPROVED** [review]
LGTM. Though, I would highly encourage you to fix all three comments.

tenor-245666676

File: cli/cmd/compliance_aws.go (hunk 10 above)
Comment (dmurray-lacework) [reply cli/cmd/compliance_aws.go:299]: Yeh, this is the enum value for the api report type so we allow it aswell as CIS

We will be allowing the value of with the version too, CIS_1_1

File: cli/cmd/compliance_aws.go (hunk 12 above)
Comment (dmurray-lacework) [reply cli/cmd/compliance_aws.go:283]: 👍

File: cli/cmd/compliance_aws.go (hunk 11 above)
Comment (dmurray-lacework) [reply cli/cmd/compliance_aws.go:315]: 👍

**Pull request merged**